iocore/eventsystem/UnixEventProcessor.cc

```diff
@@ -32,399 +32,229 @@
 {
-  ink_mutex_destroy(&dedicated_thread_spawn_mutex);
 }
 
-namespace
+ink_thread
+EventProcessor::ThreadGroupDescriptor::start(pthread_barrier_t &grpBarrier, EThread *&epEThread, int stacksize, const ThreadFxn_t &runFxn)
 {
-Event *
-make_event_for_scheduling(Continuation *c, int event_code, void *cookie)
-{
-  Event *e = eventAllocator.alloc();
+  int const affid = numa::new_affinity_id();
+  ink_semaphore stackWait;
 
-  e->init(c);
-  e->mutex          = c->mutex;
-  e->callback_event = event_code;
-  e->cookie         = cookie;
+  auto launchFxnRef = [this,&epEThread,affid,&stackWait,&grpBarrier](EThread *t) -> int
+  {
+    // NOTE: "grab" unique location [atomic]
+    EThread *&tgEThread = _thread[_count++];
 
-  return e;
-}
-}
+    ptrdiff_t id = ( &tgEThread - &_thread[0] ); // record index of it
 
-Event *
-EventProcessor::schedule_spawn(Continuation *c, EventType ev_type, int event_code, void *cookie)
-{
-  Event *e = make_event_for_scheduling(c, event_code, cookie);
-  ink_assert(ev_type < MAX_EVENT_TYPES);
-  thread_group[ev_type]._spawnQueue.enqueue(e);
-  return e;
-}
+    char thr_name[MAX_THREAD_NAME_LENGTH+1];
+    snprintf(thr_name, sizeof(thr_name), "[%s %ld]", _name.get(), id);
+    ink_set_thread_name(thr_name);
+    
+    tgEThread = t;
+    epEThread = t;
 
-Event *
-EventProcessor::schedule_spawn(void (*f)(EThread *), EventType ev_type)
-{
-  Event *e = make_event_for_scheduling(&Thread_Init_Func, EVENT_IMMEDIATE, reinterpret_cast<void *>(f));
-  ink_assert(ev_type < MAX_EVENT_TYPES);
-  thread_group[ev_type]._spawnQueue.enqueue(e);
-  return e;
-}
+    t->id = id; // index
+    t->set_affinity_id(affid);
 
-EventType
-EventProcessor::register_event_type(char const *name)
-{
-  ThreadGroupDescriptor *tg = &(thread_group[n_thread_groups++]);
-  ink_release_assert(n_thread_groups <= MAX_EVENT_TYPES); // check for overflow
+    // parallel with waiting parent now 
+    ink_sem_post(&stackWait); 
+    // parent is released to create next thread
+    
+    // gather up all threads before running
+    pthread_barrier_wait(&grpBarrier);
 
-  tg->_name = ats_strdup(name);
-  return n_thread_groups - 1;
-}
+    // change to new cpuset [real delay]
+    numa::assign_thread_cpuset_by_affinity(_affcfg, affid); 
 
-EventType
-EventProcessor::spawn_event_threads(char const *name, int n_threads, size_t stacksize)
-{
-  int ev_type = this->register_event_type(name);
-  this->spawn_event_threads(ev_type, n_threads, stacksize);
-  return ev_type;
-}
+    auto cpuset = numa::get_cpuset_by_affinity(_affcfg, affid); 
+    pretty_print_cpuset(thr_name, static_cast<hwloc_obj_type_t>(_affcfg), affid, cpuset);
 
-EventType
-EventProcessor::spawn_event_threads(EventType ev_type, int n_threads, size_t stacksize)
-{
-  char thr_name[MAX_THREAD_NAME_LENGTH];
-  int i;
-  ThreadGroupDescriptor *tg = &(thread_group[ev_type]);
+    // perform all inits for EThread 
+    for( auto &&cb : _spawnQueue ) {
+      cb(t); // init calls
```

Comment:
How will this work correctly if the callbacks are needed to initialize thread local storage, especially the C++ variety?